File: client/src/wots.ts
import { keccak_256 } from "@noble/hashes/sha3";

/**
 * Winternitz one-time signatures, client side.
 *
 * This is a line-by-line twin of the Rust in ../program/src. The two are kept
 * honest by a shared test vector rather than by inspection: check-vector.ts
 * signs a fixed message here and compares against the bytes the Rust tests
 * print. A hash-based scheme that disagrees with its verifier by one byte
 * fails in exactly the same way as a wrong key, so this is not optional.
 */

export const N = 24;
export const MSG_CHUNKS = 32;
export const SUM_CHUNKS = 2;
export const CHUNKS = MSG_CHUNKS + SUM_CHUNKS;
export const SIG_LEN = CHUNKS * N;
export const CHAIN_MAX = 255;

export const cat = (...parts: Uint8Array[]): Uint8Array => {
  const out = new Uint8Array(parts.reduce((n, p) => n + p.length, 0));
  let at = 0;
  for (const p of parts) {
    out.set(p, at);
    at += p.length;
  }
  return out;
};

export const utf8 = (s: string) => new TextEncoder().encode(s);
export const u64 = (v: bigint | number) => {
  const b = new Uint8Array(8);
  new DataView(b.buffer).setBigUint64(0, BigInt(v), true);
  return b;
};
export const hex = (b: Uint8Array) => Buffer.from(b).toString("hex");

/** One step along chain `index`. The index is inside the hash, so a link from
 *  one chunk cannot be replayed into another. */
function step(index: number, value: Uint8Array): Uint8Array {
  const buf = new Uint8Array(1 + N);
  buf[0] = index;
  buf.set(value, 1);
  return keccak_256(buf).slice(0, N);
}

export function walk(index: number, start: Uint8Array, times: number): Uint8Array {
  let cur = start;
  for (let i = 0; i < times; i++) cur = step(index, cur);
  return cur;
}

/** Secret for one chain, derived so the signer only keeps 32 bytes. */
export function secret(seed: Uint8Array, i: number): Uint8Array {
  return keccak_256(cat(utf8("WNTR:SK"), seed, Uint8Array.of(i))).slice(0, N);
}

export function publicKeyHash(seed: Uint8Array): Uint8Array {
  const ends = new Uint8Array(SIG_LEN);
  for (let i = 0; i < CHUNKS; i++) ends.set(walk(i, secret(seed, i), CHAIN_MAX), i * N);
  return keccak_256(ends);
}

/** Digest bytes, then the checksum that makes forging by walking forward useless. */
export function chunksOf(digest: Uint8Array): Uint8Array {
  const out = new Uint8Array(CHUNKS);
  out.set(digest, 0);
  let sum = 0;
  for (const b of digest) sum += CHAIN_MAX - b;
  out[MSG_CHUNKS] = (sum >> 8) & 0xff;
  out[MSG_CHUNKS + 1] = sum & 0xff;
  return out;
}

export function sign(seed: Uint8Array, digest: Uint8Array): Uint8Array {
  const counts = chunksOf(digest);
  const sig = new Uint8Array(SIG_LEN);
  for (let i = 0; i < CHUNKS; i++) sig.set(walk(i, secret(seed, i), counts[i]), i * N);
  return sig;
}

/** Local mirror of the on-chain check, so a failure can be told apart from a
 *  transport problem before anything is sent. */
export function verify(pkHash: Uint8Array, digest: Uint8Array, sig: Uint8Array): boolean {
  if (sig.length !== SIG_LEN) return false;
  const counts = chunksOf(digest);
  const ends = new Uint8Array(SIG_LEN);
  for (let i = 0; i < CHUNKS; i++) {
    ends.set(walk(i, sig.slice(i * N, (i + 1) * N), CHAIN_MAX - counts[i]), i * N);
  }
  return hex(keccak_256(ends)) === hex(pkHash);
}

File: program/src/wots.rs
//! The signing half of the scheme.
//!
//! This never runs on-chain: a vault only ever verifies. It lives here so the
//! program and the client are checked against the same code rather than
//! against two readings of the same description, which is how hash-based
//! schemes usually go wrong.

use crate::{CHAIN_MAX, CHUNKS, N, SIG_LEN};
use solana_program::keccak;

/// Secret for chain `i`, derived so a signer only has to keep 32 bytes.
pub fn secret(seed: &[u8; 32], i: u8) -> [u8; N] {
    let h = keccak::hashv(&[b"WNTR:SK", seed, &[i]]).0;
    let mut out = [0u8; N];
    out.copy_from_slice(&h[..N]);
    out
}

pub fn public_key_hash(seed: &[u8; 32]) -> [u8; 32] {
    let mut ends = [0u8; SIG_LEN];
    for i in 0..CHUNKS {
        let end = crate::walk(i as u8, &secret(seed, i as u8), CHAIN_MAX);
        ends[i * N..(i + 1) * N].copy_from_slice(&end);
    }
    keccak::hash(&ends).0
}

/// Reveals each chain at the position the digest asks for.
pub fn sign(seed: &[u8; 32], digest: &[u8; 32]) -> [u8; SIG_LEN] {
    let counts = crate::chunks_of(digest);
    let mut sig = [0u8; SIG_LEN];
    for i in 0..CHUNKS {
        let link = crate::walk(i as u8, &secret(seed, i as u8), counts[i]);
        sig[i * N..(i + 1) * N].copy_from_slice(&link);
    }
    sig
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::verify;

    fn digest_of(m: &[u8]) -> [u8; 32] {
        keccak::hash(m).0
    }

    #[test]
    fn signs_and_verifies() {
        let seed = [7u8; 32];
        let pk = public_key_hash(&seed);
        let d = digest_of(b"pay 1000 to alice");
        assert!(verify(&pk, &d, &sign(&seed, &d)));
    }

    #[test]
    fn rejects_a_different_message() {
        let seed = [7u8; 32];
        let pk = public_key_hash(&seed);
        let signed = digest_of(b"pay 1000 to alice");
        let claimed = digest_of(b"pay 1000 to mallory");
        assert!(!verify(&pk, &claimed, &sign(&seed, &signed)));
    }

    #[test]
    fn rejects_another_key() {
        let pk = public_key_hash(&[7u8; 32]);
        let d = digest_of(b"pay 1000 to alice");
        assert!(!verify(&pk, &d, &sign(&[8u8; 32], &d)));
    }

    /// The checksum is the part that makes a one-time signature safe against
    /// the obvious attack: take a signature, raise a message byte, and walk
    /// that chain forward for free. Raising a byte lowers the checksum, and
    /// lowering it needs a chain walked backwards, which nobody can do.
    #[test]
    fn checksum_blocks_walking_a_chunk_forward() {
        let seed = [7u8; 32];
        let pk = public_key_hash(&seed);
        let mut d = digest_of(b"pay 1 to alice");
        d[0] = 10;
        let sig = sign(&seed, &d);

        // Forge a digest whose first byte is higher, and advance that chain to
        // match. Every other chunk is copied across untouched.
        let mut forged = d;
        forged[0] = 200;
        let mut fsig = sig;
        let mut link = [0u8; N];
        link.copy_from_slice(&sig[..N]);
        let advanced = crate::walk(0, &link, 190);
        fsig[..N].copy_from_slice(&advanced);

        assert!(!verify(&pk, &forged, &fsig), "checksum failed to catch the forgery");
    }

    #[test]
    fn rejects_a_truncated_signature() {
        let seed = [7u8; 32];
        let pk = public_key_hash(&seed);
        let d = digest_of(b"pay 1000 to alice");
        let sig = sign(&seed, &d);
        assert!(!verify(&pk, &d, &sig[..SIG_LEN - 1]));
    }

    /// Printed so the TypeScript client can be checked against the exact bytes
    /// the program will verify, instead of against a second reading of the spec.
    #[test]
    fn print_test_vector() {
        let seed = [1u8; 32];
        let d = digest_of(b"winternitz test vector");
        let sig = sign(&seed, &d);
        println!("seed      : {}", hex(&seed));
        println!("pk_hash   : {}", hex(&public_key_hash(&seed)));
        println!("digest    : {}", hex(&d));
        println!("sig[0..24]: {}", hex(&sig[..N]));
        println!("sig_hash  : {}", hex(&keccak::hash(&sig).0));
    }

    fn hex(b: &[u8]) -> String {
        b.iter().map(|x| format!("{x:02x}")).collect()
    }
}
